src/util/string.cpp

```diff
@@ -1068,14 +1069,33 @@ void safe_print_string(std::ostream &os, std::string_view str)
 	os.setf(flags);
 }
 
-
-v3f str_to_v3f(std::string_view str)
+std::optional<v3f> str_to_v3f(std::string_view str)
 {
-	v3f value;
-	Strfnd f(str);
-	f.next("(");
-	value.X = stof(f.next(","));
-	value.Y = stof(f.next(","));
-	value.Z = stof(f.next(")"));
-	return value;
+    v3f value;
+
+    if (str.empty()) {
+        return std::nullopt;
+    }
+
+    // Strip parentheses if they exist
+    if (str.front() == '(' && str.back() == ')') {
+        str.remove_prefix(1);
+        str.remove_suffix(1);
+    }
+
+    // Replace commas with spaces
```

Comment:
Technically overly permissive: This will accept e.g. `,,1,2 , ,, 3<arbitrary suffix>`

Here's how I would go about this (without regex): Suppose you had a lambda `skip_delimiter` like this

```c++
const auto expect_delimiter = [&]() {
    const auto c = iss.get();
    return c == ' ' || c == ',';
};
```

(This could also be a function taking `iss` as a mutable reference, but you do really want it to be namespaced to just `str_to_v3f`.)

Then you could do

```c++
v3f value;
if (!(iss >> value.X)) return std::nullopt;
if (!expect_delimiter()) return std::nullopt;
if (!(iss >> value.Y)) return std::nullopt;
if (!expect_delimiter()) return std::nullopt;
if (!(iss >> value.Z)) return std::nullopt;
```

p.s. Writing parsers is something where you should be able to turn your brain off, because how grammars translate to recursive descent parsing is usually quite straightforward.